Istep 13 2

krystian-hebel opened the pull request.

krystian-hebel committed:
cpu/power9: fix directory structure and timer build fault

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
soc/ibm/power9: add I2C and SPD functions

Also clear Makefile.inc and other build issues.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
mainboard/talos-2: read SPD and fill DIMM attributes

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
device/dram/ddr4.c: fill missing ECC info from SPD

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
soc/power9: implement istep 13.2

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
include/cpu/power/scom.h: add RMW variants without a chiplet ID

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

IgorBagnucki commented on src/soc/ibm/power9/i2c.c:
```diff
@@ -0,0 +1,172 @@
+#define FIFO_REG(bus)		(0xA0004 | ((bus) << 12))
+#define CMD_REG(bus)		(0xA0005 | ((bus) << 12))
+#define MODE_REG(bus)		(0xA0006 | ((bus) << 12))
+#define STATUS_REG(bus)		(0xA000B | ((bus) << 12))
```
Please align the defines for better readability

krystian-hebel replied:
I'm not sure why GitHub displays it this way, this is properly aligned in the code. I changed the alignment of the following defines, there are now indented according to coreboot's coding style (8 chars wide).

IgorBagnucki commented on src/soc/ibm/power9/i2c.c:
```diff
@@ -0,0 +1,172 @@
+			return -1;
+		}
+
+		rnw = (segment[i].flags & I2C_M_RD) ? READ_NOT_WRITE : 0;
```
I would change the variable name. It is not very meaningful at first sight

krystian-hebel replied:
Changed.

IgorBagnucki commented on src/soc/ibm/power9/istep_13_2.c:
```diff
@@ -0,0 +1,158 @@
+void istep_13_2(dimm_attr dimms[CONFIG_DIMM_MAX])
+{
+	printk(BIOS_EMERG, "starting istep 13.2\n");
+	int d;
```
Please change the variable name to something more meaningful.

krystian-hebel replied:
Changed.

krystian-hebel committed:
soc/ibm/power9/i2c.c: fixes after review

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

miczyg1 commented on src/include/cpu/power/istep_13.h:
```diff
@@ -0,0 +1,5 @@
+
+#include <device/dram/ddr4.h>
+
+void istep_13_2(dimm_attr dimms[CONFIG_DIMM_MAX]);
```
```suggestion
void istep_13_2(dimm_attr *dimms);
```

Wouldn't be better? Otherwise, DIMM_MAX would change the function declaration

krystian-hebel replied:
That was the point, the compiler should warn if old object files are trying to be linked with newer ones. This still shouldn't happen with properly set Makefile/Kconfig. Function gets the pointer as an argument anyway, this doesn't change. What changes is that it enables some additional compile-time boundary checks and possibly optimizations. Number of DIMM slots doesn't change from one boot to another so I don't see why we need to support unknown number of elements in the array.

krystian-hebel committed:
soc/power9/istep_13_2.c: fix logic

Hostboot uses "for each MC". This __probably__ means "for each MCS", and
not "for each MCBIST" as was assumed earlier.

If that is not the case revert this commit.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

macpijan commented:
@miczyg1 Do you have any more comments to this MR?

miczyg1 commented on src/mainboard/raptor-cs/talos-2/romstage.c:
```diff
@@ -3,12 +3,40 @@
+#include <lib.h>	// hexdump
+#include <spd_bin.h>
 
 void main(void)
```
At this point this function should be called mainboard_get_spd since we do nothing mainboard-specific besides SPD.

The main should be moved to src/soc/ibm/power9/romstage.c or something like that with all the step calls and SPD decoding.

miczyg1 commented on src/soc/ibm/power9/istep_13_2.c:
```diff
@@ -0,0 +1,180 @@
+		  [all] 0
+		  [1]   PCB_EP_RESET =  1
+		*/
+		write_scom_for_chiplet(mcs_ids[i], 0x070F0042, PPC_BIT(1));
```
Let's make defines for registers instead of pasting the hex values. It should increase the readability, and reduce the number of lines of the comments. The same goes for PCB_EP_RESET and other bits used here

miczyg1 reviewed (changes requested):
Looks good besides the code arrangement and readability

macpijan commented:
Already integrated in: https://github.com/3mdeb/coreboot/pull/49
